scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetReadFn.scala

```diff
@@ -0,0 +1,237 @@
+import org.apache.parquet.{HadoopReadOptions, ParquetReadOptions}
+import org.slf4j.LoggerFactory
+
+import java.util.{Set => JSet}
+import scala.jdk.CollectionConverters._
+import scala.util.Try
+
+class ParquetReadFn[T, R](
+  readSupportFactory: ReadSupportFactory[T],
+  conf: SerializableConfiguration,
+  projectionFn: SerializableFunction[T, R]
+) extends DoFn[ReadableFile, R] {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val SPLIT_LIMIT = 64000000L
+
+  private def parquetFileReader(file: ReadableFile): ParquetFileReader = {
+    val options = HadoopReadOptions.builder(conf.get()).build
+    ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+  }
+
+  @GetRestrictionCoder def getRestrictionCoder = new OffsetRange.Coder
+
+  @GetInitialRestriction
+  def getInitialRestriction(@Element file: ReadableFile): OffsetRange = {
+    val reader = parquetFileReader(file)
+
+    val offsetRangeResult = Try(new OffsetRange(0, reader.getRowGroups.size)).toEither
+    reader.close()
+
+    offsetRangeResult.fold(throw _, identity)
+  }
+
+  @NewTracker def newTracker(@Restriction restriction: OffsetRange) =
+    new OffsetRangeTracker(restriction)
+
+  @GetSize
+  def getSize(@Element file: ReadableFile, @Restriction restriction: OffsetRange): Double = {
+    val (_, size) = getRecordCountAndSize(file, restriction)
+    size
+  }
+
+  @SplitRestriction
+  def split(
+    @Restriction restriction: OffsetRange,
+    output: DoFn.OutputReceiver[OffsetRange],
+    @Element file: ReadableFile
+  ): Unit = {
+    val reader = parquetFileReader(file)
+
+    val splitResult = Try {
+      splitRowGroupsWithLimit(
+        restriction.getFrom,
+        restriction.getTo,
+        reader.getRowGroups.asScala.toSeq
+      ).foreach(output.output)
+    }.toEither
+
+    reader.close()
+
+    splitResult.fold(throw _, identity)
+  }
+
+  @ProcessElement
+  def processElement(
+    @Element file: ReadableFile,
+    tracker: RestrictionTracker[OffsetRange, Long],
+    outputReceiver: DoFn.OutputReceiver[R]
+  ): Unit = {
+    logger.debug(
+      "start {} to {}",
+      tracker.currentRestriction.getFrom,
+      tracker.currentRestriction.getTo
+    )
+    val options: ParquetReadOptions = HadoopReadOptions.builder(conf.get()).build
+
+    val reader = ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+    try {
+      val filter = options.getRecordFilter
+      val hadoopConf = options.asInstanceOf[HadoopReadOptions].getConf
+      val parquetFileMetadata = reader.getFooter.getFileMetaData
+      val fileSchema = parquetFileMetadata.getSchema
+      val fileMetadata = parquetFileMetadata.getKeyValueMetaData
+      val readSupport = readSupportFactory.readSupport
+
+      val readContext = readSupport.init(
+        new InitContext(
+          hadoopConf,
+          fileMetadata.asScala.map { case (k, v) =>
+            k -> ImmutableSet.of(v).asInstanceOf[JSet[String]]
```

Comment:
Why do you need a cast here ?

Reply:
Would adding a type ascription work ?
```suggestion
            k -> (ImmutableSet.of(v): JSet[String])
```